### ISSUE TITLE
Container exit does not clear IP occupation

### ISSUE THREAD

**Issue by junnplus** [containerd/nerdctl#1259]
### Description

nerdctl
```
$ sudo nerdctl network create test-br
61c84410830227bcdb72d9c2bb08aec945d85f5c69d35600bfcb24168464a404

$ sudo nerdctl run --net test-br alpine ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0@if968: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 0e:ad:43:ab:24:2d brd ff:ff:ff:ff:ff:ff
    inet 10.4.6.2/24 brd 10.4.6.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::cad:43ff:feab:242d/64 scope link
       valid_lft forever preferred_lft forever

$ sudo nerdctl run --net test-br alpine ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0@if969: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 52:5e:96:74:c9:a6 brd ff:ff:ff:ff:ff:ff
    inet 10.4.6.3/24 brd 10.4.6.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::505e:96ff:fe74:c9a6/64 scope link
       valid_lft forever preferred_lft forever
```

docker 
```
$ docker network create test-br
d1d9b84082c5ffff180ea111767412d06d6690d5015602421696075ede7cdf26

$ docker run --net test-br alpine ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
963: eth0@if964: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:c0:a8:20:02 brd ff:ff:ff:ff:ff:ff
    inet 192.168.32.2/20 brd 192.168.47.255 scope global eth0
       valid_lft forever preferred_lft forever

$ docker run --net test-br alpine ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
965: eth0@if966: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:c0:a8:20:02 brd ff:ff:ff:ff:ff:ff
    inet 192.168.32.2/20 brd 192.168.47.255 scope global eth0
       valid_lft forever preferred_lft forever
```

### Steps to reproduce the issue

1.
2.
3.

### Describe the results you received and expected

keep the same as docker

### What version of nerdctl are you using?

master

**Comment by t1anz0ng** [comment]
/assign 
🙋‍♂️

**Comment by t1anz0ng** [comment]
After thinking a little about this behavior,  I'm not sure if we need  to keep the same as `docker`  

Because in nerdctl the IP is allocated/released by `host-local` `IPAM`  by
default and It actually clears the IP occupation, but just round-robin IP address.  

I created >256 (actually only need 254?) containers by 
`docker run --nettest-br alpine ip addr`,  we can see that the released  IP is available.
 
Can also observe from the `host-local`'s dataDir that the existed  container's IP is released, 
which is `/var/lib/cni/networks/test-br` in this case. 

https://github.com/containernetworking/plugins/blob/v1.1.1/plugins/ipam/host-local/backend/disk/backend.go#L149  

@junnplus

**Comment by AkihiroSuda** [comment]
Looks like this issue has to be fixed on the CNI plugin side (if we still want to fix it?)

**Comment by junnplus** [comment]
```sh
$ sudo nerdctl network create test-br
61c84410830227bcdb72d9c2bb08aec945d85f5c69d35600bfcb24168464a404
$ sudo nerdctl run --net test-br alpine ip addr
2: eth0@if38: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether ee:38:e6:86:49:90 brd ff:ff:ff:ff:ff:ff
    inet 10.4.1.2/24 brd 10.4.1.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::ec38:e6ff:fe86:4990/64 scope link
       valid_lft forever preferred_lft forever
$ sudo nerdctl run --net test-br --ip=10.4.1.2 alpine ip addr
FATA[0000] failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: 
error during container init: error running hook #0: error running hook: exit status 1, stdout: , 
stderr: time="2022-08-03T13:36:01Z" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to allocate for range 0: requested IP address 10.4.1.2 is not available in range set 10.4.1.1-10.4.1.254"
Failed to write to log, write /var/lib/nerdctl/1935db59/containers/default/f5a6e2556ebbf6af11f06b4d35bd3541fa7d8a6463a59c956649e7c70edcb5ba/oci-hook.createRuntime.log: file already closed: unknown
```

**Comment by t1anz0ng** [comment]
> I haven't confirmed the implementation of host-local, but in fact the IP doesn't seem to be released

It will release after `nerdctl rm <cont_id>`.

TLDR:
`nerdctl run --rm` will release after container exit.

```
# nerdctl run --net test-br --rm alpine ip addr
...
2: eth0@if332: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether b6:9b:55:32:33:16 brd ff:ff:ff:ff:ff:ff
    inet 10.4.2.11/24 brd 10.4.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::b49b:55ff:fe32:3316/64 scope link
       valid_lft forever preferred_lft forever

root in cni/networks/test-br took 2s
❯ nerdctl run --net test-br --ip=10.4.2.11 alpine ip addr
...
2: eth0@if333: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 1e:e3:e4:3b:f3:50 brd ff:ff:ff:ff:ff:ff
    inet 10.4.2.11/24 brd 10.4.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::1ce3:e4ff:fe3b:f350/64 scope link
       valid_lft forever preferred_lft forever
```

more detailed process:
```
# sudo nerdctl run --net test-br --name=test alpine ip addr
...
    inet 10.4.2.10/24 brd 10.4.2.255 scope global eth0
...

# ls /var/lib/cni/networks/test-br
10.4.2.10  last_reserved_ip.0  lock

# cat /var/lib/cni/networks/test-br/last_reserved_ip.0
10.4.2.10

# sudo nerdctl run --net test-br --ip=10.4.2.10 alpine ip addr
FATA[0001] failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2022-08-03T23:51:18+08:00" level=fatal msg="failed to call cni.Setup: plugin type=\"bridge\" failed (add): failed to allocate for range 0: requested IP address 10.4.2.10 is not available in range set 10.4.2.1-10.4.2.254"
Failed to write to log, write /var/lib/nerdctl/1935db59/containers/default/3c6e967945daefb0d6bac9fbad418fa5e00364f071258cd79094853080c0ba84/oci-hook.createRuntime.log: file already closed: unknown

#  sudo nerdctl rm test
test

# nerdctl run --net test-br --ip=10.4.2.10 alpine ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0@if331: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 46:25:d6:34:12:db brd ff:ff:ff:ff:ff:ff
    inet 10.4.2.10/24 brd 10.4.2.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::4425:d6ff:fe34:12db/64 scope link
       valid_lft forever preferred_lft forever

```

**Comment by junnplus** [comment]
> It will release after nerdctl rm <cont_id>

I know that deleting the container can release the IP, I mean stop a container.

**Comment by obataku** [comment]
@junnplus did it formerly release IP on container exit? i.e. is this a breaking change in behavior or has it always been like this, to your knowledge?

**Comment by obataku** [comment]
@junnplus did it formerly release IP on container exit? i.e. is this a breaking change in behavior or has it always been like this, to your knowledge?

**Comment by junnplus** [comment]
@obataku it always been like this

**Comment by zhan9san** [comment]
related to #458
